Fix: Menu anchor respect alignment offset for right

rkishan516 opened the pull request:
Fix: Menu anchor respect alignment offset for right
fixes: #159413 

Before this PR :-

https://github.com/user-attachments/assets/644ec10f-c1c6-454d-a709-b98abfc3af05

After this PR :-

https://github.com/user-attachments/assets/116dfad5-1ca8-44f7-81f5-f39ca30d3c7d

So, change after this PR is after overflow it doesn't just stick to the wall, rather it comes other side of button and adjust the offset on other side.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

rkishan516 committed:
Fix: Menu anchor respect alignment offset for right

TahaTesser commented on packages/flutter/test/material/menu_anchor_test.dart:
```diff
@@ -3724,6 +3724,53 @@ void main() {
+      await changeSurfaceSize(tester, const Size(800, 600));
+      await tester.pumpWidget(
+        MaterialApp(
+          theme: ThemeData(useMaterial3: false),
```
This can be omitted if this is not specific to M2.

rkishan516 replied:
I wanted to be consistent with other tests, specifically alignment test of menu anchor.

TahaTesser commented on packages/flutter/test/material/menu_anchor_test.dart:
```diff
@@ -3724,6 +3724,53 @@ void main() {
+          home: Builder(
+            builder: (BuildContext context) {
+              return Directionality(
+                textDirection: TextDirection.ltr,
```
Directionality is provided by `MaterialApp`, this can be skipped.

rkishan516 replied:
Sure

TahaTesser commented on packages/flutter/test/material/menu_anchor_test.dart:
```diff
@@ -3724,6 +3724,53 @@ void main() {
+      expect(find.byType(MenuItemButton), findsOneWidget);
+      expect(
+        collectSubmenuRects(),
+        equals(const <Rect>[Rect.fromLTRB(634.0, 488.0, 756.0, 552.0)]),
```
Could please add a comment on how these numbers verify "menus should respect overflow on right with the provided offset"?

rkishan516 replied:
Sure

TahaTesser reviewed (changes requested):
Running test code without this PR and with this PR has no difference. In fact, the PR test also passes without the fix. 

### Code Sample

<details>
<summary>expand to view the code sample</summary> 

```dart
import 'package:flutter/material.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(useMaterial3: false),
      home: Align(
        alignment: Alignment.bottomRight,
        child: MenuAnchor(
          alignmentOffset: const Offset(-50, 0),
          menuChildren: const <Widget>[MenuItemButton(child: Text('Button1'))],
          builder: (BuildContext context, MenuController controller, Widget? child) {
            return FilledButton(
              onPressed: () {
                if (controller.isOpen) {
                  controller.close();
                } else {
                  controller.open();
                }
              },
              child: const Text('Tap me'),
            );
          },
        ),
      ),
    );
  }
}

```

</details>

<img width="650" alt="Screenshot 2025-03-12 at 16 38 53" src="https://github.com/user-attachments/assets/358c1f17-c334-4ce0-b02f-d06cca3e077f" />


Please make sure the fix is effective. Update PR description with before and after to showcase the fix visually.

rkishan516 commented:
> Running test code without this PR and with this PR has no difference. In fact, the PR test also passes without the fix.
> 
> ### Code Sample
> expand to view the code sample
> <img alt="Screenshot 2025-03-12 at 16 38 53" width="650" src="https://private-user-images.githubusercontent.com/48603081/421939712-358c1f17-c334-4ce0-b02f-d06cca3e077f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE3OTE2MzksIm5iZiI6MTc0MTc5MTMzOSwicGF0aCI6Ii80ODYwMzA4MS80MjE5Mzk3MTItMzU4YzFmMTctYzMzNC00Y2UwLWIwMmYtZDA2Y2NhM2UwNzdmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzEyVDE0NTUzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdhMWU2OGU4MTE4NzM5ODAxNDRiODEzZDc3ODBjYTI5NjVmMzA4NTc4YzYwZjA2NjEzMTVlNjY2NjkyM2E5OGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.4ET8aR6Rjtd5id1Y90vPNXJezvpVT8Be7r03uIBmoiU">
> Please make sure the fix is effective. Update PR description with before and after to showcase the fix visually.

Ohh, I think this need to have one change to see effect that offset should be positive.
Also, for both M2 and M3 result is going to be same, so i will remove material2 condition.

dkwingsmt commented:
Gentle nudge @TahaTesser to review, or if you'd like someone else to take over. :)  (from triage)

TahaTesser commented on packages/flutter/test/material/menu_anchor_test.dart:
```diff
@@ -3724,6 +3724,49 @@ void main() {
+      await tester.pump();
+
+      expect(find.byType(MenuItemButton), findsOneWidget);
+      // Initial position without offset in x (667.4,790.1).
+      // Ending position with offset 50 should be in x (717.4,840.1).
+      // So, instread of right it moves to left anchor of button and placed
+      // at (667.4 - offset - width of menu, 667.4 - offset) i.e. (494.7, 617.4).
+      expect(
+        collectSubmenuRects(),
+        equals(const <Rect>[Rect.fromLTRB(494.7000045776367, 488.0, 617.4000015258789, 552.0)]),
```
It still not clear how we get `(667.4,790.1)` and  `(717.4,840.1)` in the comment. Generally, it is better to calculate relative position of a button than absolute numbers like these. For instance calculating where the button is rendered relative to the window or layout position.

rkishan516 replied:
Thanks for suggestion, might be comment can be written in better way as you said. So, here `(667.4,790.1)` is the starting and ending position in x, not coordinate. Which then offset by 50 changes to `(717.4,840.1)`. But yes I agree this can be written in a better way as you suggested.

TahaTesser reviewed (changes requested):
Thanks for attempting to fix this. I'm still confused how the fix is tested.  

I see the button is positioned to the end like this without the fix.
<img width="798" alt="Screenshot 2025-03-28 at 21 59 20" src="https://github.com/user-attachments/assets/a07bffb5-8324-46e1-a1a1-d0112e278e41" />

Then with the fix, the button is positioned way off.  Is this correct expectation? 
<img width="798" alt="Screenshot 2025-03-28 at 21 59 00" src="https://github.com/user-attachments/assets/8763c75d-6eb1-4968-ba23-3d06ce96643e" />

As I mentioned earlier in https://github.com/flutter/flutter/pull/164858#pullrequestreview-2678579590, 

>  Update PR description with before and after to showcase the fix visually.
 
Could you please update PR description with more details and explain visually how this PR affects the button position?

rkishan516 commented:
> Thanks for attempting to fix this. I'm still confused how the fix is tested.
> 
> I see the button is positioned to the end like this without the fix. <img alt="Screenshot 2025-03-28 at 21 59 20" width="798" src="https://private-user-images.githubusercontent.com/48603081/428178917-a07bffb5-8324-46e1-a1a1-d0112e278e41.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDMyMTE3NjYsIm5iZiI6MTc0MzIxMTQ2NiwicGF0aCI6Ii80ODYwMzA4MS80MjgxNzg5MTctYTA3YmZmYjUtODMyNC00NmUxLWExYTEtZDAxMTJlMjc4ZTQxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzI5VDAxMjQyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRkY2E3ZDdlODYwNTA1NmQ1NDA1YWFiZTU5MWZkNTk1ZWViNWY2ZGIzNWZiZWE4NTFkNmUyMDhlYzkxZmU5MDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.CFN3RecCgyFmGecr4TLpf2U3J59gfBhJWMvnRHnEHBE">
> 
> Then with the fix, the button is positioned way off. Is this correct expectation? <img alt="Screenshot 2025-03-28 at 21 59 00" width="798" src="https://private-user-images.githubusercontent.com/48603081/428179415-8763c75d-6eb1-4968-ba23-3d06ce96643e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDMyMTE3NjYsIm5iZiI6MTc0MzIxMTQ2NiwicGF0aCI6Ii80ODYwMzA4MS80MjgxNzk0MTUtODc2M2M3NWQtNmViMS00OTY4LWJhMjMtM2QwNmNlOTY2NDNlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzI5VDAxMjQyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ4NzNjMjg2ZDI5ZDhiMzE4ZDJlN2NhNGExZjhhMDI3MGE4ZWYzOWFhZTU2NzlhMWE4YmQ0OGMyNjcwY2ZlZjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.xbslurGA1_ZGB4cohrAR3NFe46oYYaTaLa1W2qqRiJo">
> 
> As I mentioned earlier in [#164858 (review)](https://github.com/flutter/flutter/pull/164858#pullrequestreview-2678579590),
> 
> > Update PR description with before and after to showcase the fix visually.
> 
> Could you please update PR description with more details and explain visually how this PR affects the button position?

Yes, current implementation is like this, where it will try to fit bottomLeft until it can, if it goes out of screen than that same offset applies on other side of button. And that is the reason you are seeing this.

rkishan516 commented:
> Could you please update PR description with more details and explain visually how this PR affects the button position?

Sure, I will update PR description.

rkishan516 commented:
@TahaTesser I have cleaned up test to look simple. Also I have updated PR description with visual feedback that shows how with alignment offset, position of submenu is changing.
